tests/pubsub_logging_test.py

```diff
@@ -32,8 +39,126 @@ class CompatBase64Test(unittest.TestCase):
+        self.r = logging.LogRecord('test', logging.INFO, None, 0, self.log_msg,
+                                   [], None)
+
+    def publish(self):
+        publish_body(self.mocked_client, self.expected_body, self.topic,
+                     self.RETRY)
+
+    def test_publish_body(self):
+        """Basic test for publish_body."""
+        self.publish()
+        self.topics.publish.assert_called_once_with(
+            topic=self.topic, body=self.expected_body)
+        self.topics_publish.execute.assert_called_with(num_retries=self.RETRY)
+
+    def test_publish_body_raise_on_publish_404(self):
+        """Tests if the flush method raises when publish gets a 404 error."""
+        mocked_resp = mock.MagicMock()
+        mocked_resp.status = 404
+        mocked_resp.reason = 'Not Found'
+        # 404 error
+        self.topics_publish.execute.side_effect = [
+            errors.HttpError(mocked_resp, 'Not found')
+        ]
+        self.assertRaises(errors.HttpError, self.publish)
+
+    def test_flush_raise_on_publish_403(self):
+        """Tests if the flush method raises when publish gets a 403 error."""
+        mocked_resp = mock.MagicMock()
+        mocked_resp.status = 403
+        mocked_resp.reason = 'Access not allowed'
+        # 403 error
+        self.topics_publish.execute.side_effect = [
+            errors.HttpError(mocked_resp, 'Access not allowed'),
+        ]
+        self.assertRaises(errors.HttpError, self.publish)
+
+    def test_flush_ignore_recoverable(self):
+        """Tests if we raise upon getting 503 error from Cloud Pub/Sub."""
+        mocked_resp = mock.MagicMock()
+        mocked_resp.status = 503
+        mocked_resp.reason = 'Server Error'
+        # 503 error
+        self.topics_publish.execute.side_effect = [
+            errors.HttpError(mocked_resp, 'Server Error'),
+        ]
+        self.assertRaises(RecoverableError, self.publish)
+        self.topics.publish.assert_called_once_with(
+            topic=self.topic, body=self.expected_body)
+        self.topics_publish.execute.assert_called_once_with(
+            num_retries=self.RETRY)
+
+
+class CountPublishBody(object):
+    """A simple counter that counts total number of messages."""
+    def __init__(self):
+        self.cnt = mp.Value('i', 0)
+        self.lock = mp.Lock()
+
+    def __call__(self, client, body, topic, retry):
+        with self.lock:
+            self.cnt.value += len(body['messages'])
+
+
+class AsyncPubsubHandlerTest(unittest.TestCase):
+    """Tests for async_handler.AsyncPubsubHandler."""
+    RETRY = 10
+
+    def setUp(self):
+        self.mocked_client = mock.MagicMock()
+        self.topic = 'projects/test-project/topics/test-topic'
+        self.counter = CountPublishBody()
+        self.handler = pubsub_logging.AsyncPubsubHandler(
+            topic=self.topic, client=self.mocked_client, retry=self.RETRY,
+            worker_size=10, timeout=0.1, publish_body=self.counter)
```

Comment:
for this test, could you just patch the publish_body attr on the handler - and then avoid having to pollute your __init__ args on the handler class?

`self.handler.publish_body = self.counter`

It feels odd to have an arg to the constructor with "for testing purposes" in the docs